### PR TITLE
ENH: adds auto_close as a simulation parameter

### PR DESCRIPTION
**Pull request by warren-oneill**
Allows users to turn off automatic closing of futures by setting an ```auto_close``` flag to ```False``` in the simulation parameters.
cc @jfkirk @StewartDouglas @yankees714

### CODE REVIEW

File: tests/test_pickle_serialization.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 from nose_parameterized import parameterized
 from unittest import TestCase
 
-from .serialization_cases import (
+from tests.serialization_cases import (
```
Comment (warren-oneill) [inline tests/test_pickle_serialization.py]: I was having problems in PyCharm with relative imports but I can revert this if needs be.

**Comment by jfkirk** [comment]
@ssanderson What do you think? It makes sense to me to make auto-closing configurable, but I know that we are trying to wean ourselves off of the SimulationParameters.

**Comment by warren-oneill** [comment]
@jfkirk @ssanderson another option would be to add the flag to the asset metadata. In this way we can have algos where some assets are auto closed and others not.

**Comment by ssanderson** [comment]
Changes that modify global simulation behavior like this make me nervous; each individual such change is innocuous, but over time they create exponentially-many possible configurations in which Zipline could be running, some of which might be mutually-exclusive, and almost all of which will be untested, because we write roughly linearly-many tests, but produce exponentially-many configurations.

@warren-oneill can you talk about about why a user would or wouldn't want to auto-close positions on a contract?  I'm wondering if, for example, the right way to do this would be to add `close_date` as an optional parameter to `order`, which would default to the requested asset's expiration.

**Comment by warren-oneill** [comment]
@ssanderson for us they are a couple of cases where we would want to leave the positions open after expiration day:
* as an energy producer we can use the futures to hedge against our production
* we trade also in financial futures where there is no danger of going into delivery but instead after expiration the open position are closed using a settlement price and we may want to bet on this price
* market outages are very common in our field and there may be situations where we are unable to close our position. This is something which we would like to have in our backtests.
* some products in our markets are not as liquid as stocks and it can be that one is unable to find a buyer/seller.

**Comment by ssanderson** [comment]
@warren-oneill thoughts on an alternative proposed at https://github.com/quantopian/zipline/pull/676/files#r36980067 ?  I think baking this logic into the data model for Futures would make this logic much simpler rather than having to have `if auto_close_futures` branching everywhere, we just have `if now > future.auto_close_date`, which means we'd be exercising the same code paths at all times, which means fewer opportunities for bugs to hide.

The downside of baking this into the data model is that it'd be harder to toggle back and forth between auto_close=True and auto_close=False.  The upside would be that you'd have per-asset granularity on whether you want auto-closing behavior.  Thoughts on these tradeoffs?

**Comment by warren-oneill** [comment]
closed in favour of #676

**Pull request closed without merging**